src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -2151,8 +2151,17 @@ because xbuild doesn't support framework reference assemblies.
     <_AssembliesToProcess Include="@(ResolvedUserAssemblies)" />
     <_AssembliesToProcess Include="@(ResolvedAssemblies)" Condition=" '%(Filename)' != '' And '@(_JniFrameworkAssembly)' != '' " />
   </ItemGroup>
+  <PropertyGroup>
+    <_Is64BitOperatingSystem>$([System.Environment]::Is64BitOperatingSystem)</_Is64BitOperatingSystem>
+    <_WindowsFlavor Condition=" '$(_Is64BitOperatingSystem)' == 'true' ">Win64</_WindowsFlavor>
+    <_WindowsFlavor Condition=" '$(_Is64BitOperatingSystem)' == 'false' ">Win32</_WindowsFlavor>
+    <_MonoBinary Condition=" '$(OS)' == 'Windows_NT' ">$(MonoAndroidToolsDirectory)\lib\host-mxe-$(_WindowsFlavor)\monow.exe</_MonoBinary>
+    <_MonoBinary Condition=" '$(OS)' != 'Windows_NT' ">$(MonoAndroidBinDirectory)\mono</_MonoBinary>
+    <_MonoPath Condition=" '$(OS)' == 'Windows_NT' ">( set MONO_PATH=$(_XATargetFrameworkDirectories);$(MonoAndroidLinkerInputDir) ) &amp;<_MonoPath>
+    <_MonoPath Condition=" '$(OS)' != 'Windows_NT' ">MONO_PATH=&quot;$(_XATargetFrameworkDirectories):$(MonoAndroidLinkerInputDir)&quot;<_MonoPath>
+  </PropertyGroup>
   <Exec
-       Command="MONO_PATH=&quot;$(_XATargetFrameworkDirectories):$(MonoAndroidLinkerInputDir)&quot; &quot;$(MonoAndroidBinDirectory)\mono&quot; --debug &quot;$(MonoAndroidToolsDirectory)\jnimarshalmethod-gen.exe&quot; --jvm=&quot;$(JdkJvmPath)&quot; $(AndroidGenerateJniMarshalMethodsAdditionalArguments) @(_AssembliesToProcess->'&quot;$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)&quot;', ' ')"
+       Command="$(_MonoPath) &quot;$(_MonoBinary)&quot; --debug &quot;$(MonoAndroidToolsDirectory)\jnimarshalmethod-gen.exe&quot; --jvm=&quot;$(JdkJvmPath)&quot; $(AndroidGenerateJniMarshalMethodsAdditionalArguments) @(_AssembliesToProcess->'&quot;$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)&quot;', ' ')"
   />
```

Comment:
Just a thought, but you could make a new `JniMarshalMethodGen` MSBuild task that inherits from [ToolTask](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.utilities.tooltask?view=netframework-4.7.2). It might a lot cleaner to do all this logic in C# that checks for Windows, 32-bit, 64-bit, etc. There is also an `EnvironmentVariables` property for the new process, where you could set `MONO_PATH`.

Reply:
Yes, it was flowing in that direction for some time. Today I decided to finally do it, so that I can try different quotation on Windows to see whether it would help with the above issue.

It didn't help, but at least we have the task now, and so we have the OS specific code hidden in the task from `targets` file.